### PR TITLE
Rendering and notification field

### PR DESCRIPTION
**Pull request by tboulogne**
Hello, 

Thanks for your works. With this pull, i'd like to contribute with :
- switch to bootstrap 4
- two new fields (notification_send, ticket_raised) to avoid multiple notifications or tickets for same error
- prettify of request_data
image
- new rendering of details and list (bootstrap 4)
image
- two new blocks : head_script and header_block in template

Feel free to accept what you want.

Regards

Thierry

### CODE REVIEW

**Comment by sonus21** [comment]
Just a high level comment from me,
We should take notification send as an additional configuration, some users might want to send notifications for subsequent failures as well but other might not be like you. 

Even when we're not sending notifications, we should consider the time difference between two consecutive failures. If we're seeing time difference is greater than some delta than we should send notification, reason for this is we keep overriding the failure, that does not mean yesterday failure is same as today's failure, it could be as well but we're not sure only developer has the right knowledge about such things.

**Comment by tboulogne** [comment]
Hello

I will modify this and resend request ;-)

Thierry BOULOGNE
06 43 18 31 19

**Comment by tboulogne** [comment]
@sonus21 add parameter for notification and ticket.

File: error_tracker/django/migrations/0003_auto_20201018_1317.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Generated by Django 3.1.2 on 2020-10-18 11:17
```
Comment (sonus21) [inline error_tracker/django/migrations/0003_auto_20201018_1317.py]: We don't need this file

File: error_tracker/django/migrations/0002_auto_20201018_1311.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('django', '0001_initial'),
+    ]
+
+    operations = [
+        # migrations.AddField(
```
Comment (sonus21) [inline error_tracker/django/migrations/0002_auto_20201018_1311.py]: We would need  to add migrations for exceptions table. And please remove alter for request_data field.

File: error_tracker/django/utils.py [hunk 3]
```diff
@@ -54,15 +54,37 @@ def _get_form_data(request):
...
     @staticmethod
     def _get_headers(request):
+        # lamda fonction for cookie cleaning
+        clean_value = lambda x : x.value.replace('[["', "").replace('"]]', "").replace('"', "")
+        
         if request is not None:
             try:
-                headers = request.headers.dict()
-            except AttributeError:
+                headers = request.headers
```
Comment (sonus21) [inline error_tracker/django/utils.py]: maybe you can pull this to another method to pull cookie.

**Comment by tboulogne** [comment]
@sonus21 let me know if better now.

File: error_tracker/libs/mixins.py [hunk 4]
```diff
@@ -39,6 +39,8 @@ class ModelMixin(object):
     count = None
     created_on = None
     last_seen = None
+    notification_send = None
+    ticket_raised = None
```
Comment (sonus21) [inline error_tracker/libs/mixins.py]: Default value can be false.

File: error_tracker/django/utils.py (hunk 3 above)
Comment (tboulogne) [reply error_tracker/django/utils.py]: @sonus21 don't understand your meaning, could you explain a little bit please ?

Comment (sonus21) [reply error_tracker/django/utils.py]: You can add a function that will get cookie from request object.

**Comment by tboulogne** [comment]
@sonus21 let me know if better now please.

File: error_tracker/django/models.py [hunk 5]
```diff
@@ -30,7 +30,8 @@ class ErrorModel(models.Model, ModelMixin):
     count = models.IntegerField(default=0)
     created_on = models.DateTimeField(auto_now=True)
     last_seen = models.DateTimeField(auto_now=True, db_index=True)
-
+    notification_send = models.BooleanField(default=False)
```
Comment (sonus21) [inline error_tracker/django/models.py]: Please rename this to `notification_sent`

File: error_tracker/django/migrations/0002_auto_20201018_1311.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('django', '0001_initial'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='errormodel',
+            name='notification_send',
```
Comment (sonus21) [inline error_tracker/django/migrations/0002_auto_20201018_1311.py]: `notification_sent` as column name

File: error_tracker/django/settings.py [hunk 7]
```diff
@@ -44,3 +44,8 @@ def get(key, default):
 APP_ERROR_DB_MODEL = get('APP_ERROR_DB_MODEL', None)
 # Check error views are visible to others or not
 APP_ERROR_VIEW_PERMISSION = get('APP_ERROR_VIEW_PERMISSION', None)
+# Send email notification once
+APP_ERROR_NOTIFICATION_ONCE = get('APP_ERROR_NOTIFICATION_ONCE', None)
```
Comment (sonus21) [inline error_tracker/django/settings.py]: Here default should be False

File: error_tracker/django/settings.py [hunk 8]
```diff
@@ -44,3 +44,8 @@ def get(key, default):
...
+# Raise ticket once
+APP_ERROR_TICKET_ONCE = get('APP_ERROR_NOTIFICATION_ONCE', None)
```
Comment (sonus21) [inline error_tracker/django/settings.py]: same

File: error_tracker/django/utils.py [hunk 9]
```diff
@@ -182,3 +183,34 @@ def capture_exception(request=None, exception=None, additional_context=None):
...
+    """
+    Parse request headers to extract cookie.
+    :param headers (request headers])
+    :return: [dict]: return parse header with cookie as dict
+    """
+    new_headers = {}
+    for key, value in headers.items():
+        try:
+            # Test if value could be json loaded, parse if needed as for cookie.
+            json.loads('{"%s":"%s"}' % (key, value))
```
Comment (sonus21) [inline error_tracker/django/utils.py:203]: No usage of ` json.loads`?

**Review by sonus21 — CHANGES_REQUESTED** [review]
Overall looks good, there're just few nitpicks, once addressed I'll approve and release a new version.

File: error_tracker/django/utils.py (hunk 9 above)
Comment (tboulogne) [reply error_tracker/django/utils.py:203]: Emulate json.loads to detect problematic values and set them to ```value=""```

File: error_tracker/django/utils.py [hunk 10]
```diff
@@ -182,3 +183,34 @@ def capture_exception(request=None, exception=None, additional_context=None):
...
     error_tracker.capture_exception(request=request, exception=exception,
                                     additional_context=additional_context)
+
+
+def clean_value(x):
+    x = x.value.replace('[["', "").replace('"]]', "").replace('"', "")
+    return x
+
+
+def cookie_parse(headers):
```
Comment (sonus21) [inline error_tracker/django/utils.py]: One last change request, this method does not seem to parse only cookie, it's parsing header. Can we rename this method to `parse_headers`

File: error_tracker/django/utils.py (hunk 9 above)
Comment (sonus21) [reply error_tracker/django/utils.py:203]: Can you please provide some more details how it helps you and on what value? As you're encoding value in string something like this

```
{
   "foo" : "<value>"
}
```

As your code, any value could be parsed.

**Comment by tboulogne** [comment]
``` json.loads('{"%s":"%s"}' % (key, value))```

Parse each key, value from headers items and Test if could be "json loaded". If not, we set the correspondant value to empty except for cookie key.

because, if could not load a key, value with json, pretty display will not works... 

hope it help you to understand...

Regards

**Comment by sonus21** [comment]
> ` json.loads('{"%s":"%s"}' % (key, value))`
> 
> Parse each key, value from headers items and Test if could be "json loaded". If not, we set the correspondant value to empty except for cookie key.
> 
> because, if could not load a key, value with json, pretty display will not works...
> 

How are we going to handle form data? I'm still not convinced this will be able to catch the json load error. Can you please share some example request data?

**Comment by tboulogne** [comment]
@sonus21 could you trust me on this point :-) ? I faced the problem to provide this **tortuous** solution :-).

**Comment by tboulogne** [comment]
> > `json.loads('{"%s":"%s"}' % (key, value))`
> > Parse each key, value from headers items and Test if could be "json loaded". If not, we set the correspondant value to empty except for cookie key.
> > because, if could not load a key, value with json, pretty display will not works...
> 
> How are we going to handle form data? I'm still not convinced this will be able to catch the json load error. Can you please share some example request data?
Something like that
```
{'headers': {'Host': 'xxxxxxx', 'Accept': 'text/css,*/*;q=0.1', 'Connection': 'keep-alive', 'Cookie': {'sessionid': 'z027lpgvxo3pf85u6pkzsqdcs87cyojr', 'csrftoken': 'XXXXXXXXXXXXXXXXXXXXXXX', 'messages': '__json_message,0,25,Connexion avec myuser r\\u00e9ussie.],[__json_message,0,25,Connexion avec myuser r\\u00e9ussie.],[__json_message,0,25,Connexion avec myuser r\\u00e9ussie.],[__json_message,0,25,Connexion avec myuser r\\u00e9ussie.],[__json_message,0,25,Connexion avec myuser r\\u00e9ussie.],[__json_message,0,25,Connexion avec myuser r\\u00e9ussie.],[__json_message,0,25,Connexion avec myuser r\\u00e9ussie.],[__json_message,0,25,Connexion avec myuser r\\u00e9ussie.],[__json_message,0,25,Connexion avec myuser r\\u00e9ussie.],[__json_message,0,25,Connexion avec myuser r\\u00e9ussie.],[__json_message,0,25,Vous \\u00eates d\\u00e9connect\\u00e9.],[__json_message,0,25,Connexion avec myuser r\\u00e9ussie.:1kUbGk:y1uTcIVlaWEYbyJGMKZ5r5qisGggfuA4L6Pt2kflF4w'}, 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Safari/605.1.15', 'Accept-Language': 'fr-fr', 'Referer': 'http://127.0.0.1:8000/render_list/products_manager/manageproduct_list/_mp_import_status/ALL/', 'Accept-Encoding': 'gzip, deflate'}}```

**Comment by sonus21** [comment]
> @sonus21 could you trust me on this point :-) ? I faced the problem to provide this **tortuous** solution :-).

Approved still I need to see.

**Pull request merged**